spec/support/fake_ansible_repo.rb

```diff
@@ -0,0 +1,302 @@
+      # otherwise it will be a empty file.
+      #
+      def build_file(repo_rel_path, entry)
+        full_path = repo_rel_path.join(entry)
+        content   = if filepath_match?(full_path, "**/requirements.{yml,yaml}")
+                      REQUIREMENTS
+                    elsif filepath_match?(full_path, *VAR_FILE_FNMATCHES)
+                      VAR_DATA
+                    elsif filepath_match?(full_path, "**/roles/*/meta/main.{yml,yaml}")
+                      META_DATA
+                    elsif filepath_match?(full_path, "**/*.{yml,yaml}")
+                      dummy_playbook_data_for(full_path.basename)
+                    end
+
+        File.write(full_path, content)
+      end
+
+      # Given a collection of glob based `File.fnmatch` strings, confirm
+      # whether any of them match the given path.
+      #
+      def filepath_match?(path, *acceptable_matches)
+        acceptable_matches.any? { |match| path.fnmatch?(match, File::FNM_EXTGLOB) }
+      end
+
+      # Init new repo at local_repo
+      #
+      #   $ cd /tmp/clone_dir/hello_world_local && git init .
+      #
+      def git_init
+        @repo  = Rugged::Repository.init_at(repo_path.to_s)
+        @index = repo.index
+      end
+
+      # Add new files to index
+      #
+      #   $ git add .
+      #
+      def git_add_all
+        index.add_all
+        index.write
+      end
+
+      # Create initial commit
+      #
+      #   $ git commit -m "Initial Commit"
+      #
+      def git_commit_initial
+        Rugged::Commit.create(
+          repo,
+          :message    => "Initial Commit",
+          :parents    => [],
+          :tree       => index.write_tree(repo),
+          :update_ref => "HEAD"
+        )
+      end
+
+      # Generates dummy playbook file content
+      #
+      # If the filename ends with ".encrypted.yml", then it will be
+      # "encrypted", in which the content is made to look like it was.
+      #
+      # Since copying the encryption scheme of ansible-vault is outside the
+      # scope of this class, it simply does enough to emulate what a encrypted
+      # file looks like, which is:
+      #
+      #   - Adding the header ("$ANSIBLE_VAULT;1.1;AES256")
+      #   - "Encrypting" the data (just double converting it to hex)
+      #   - Taking the encrypted data and converting it to 80 chars
+      #
+      def dummy_playbook_data_for(filename)
+        data = <<~PLAYBOOK_DATA
+          - name: #{filename}
+            hosts: all
+            tasks:
+              - name: #{filename} Message
+                debug:
+                  msg: "Hello World! (from #{filename})"
+        PLAYBOOK_DATA
+
+        if filename.basename.fnmatch?("*.encrypted.{yml,yaml}", File::FNM_EXTGLOB)
+          to_hex = data.unpack1("H*").unpack1("H*")
+          data   = (0..to_hex.length).step(80).to_a
+                                     .map! { |start| to_hex[start, 80] }
```

Comment:
I played around with this locally, and I think it actually creates an extra blank at then end if you have *exactly* 80 chars.  Not sure it really matters though.

Reply:
s = "abcdefghij"
n = 5
(0..s.size).step(n).to_a.map! { |start| s[start, n] }
# => ["abcde", "fghij", ""]

The simple fix seems to be to just do `(0...size)` vs `(0..size)`

s = "abcdefghij"
n = 5
(0...s.size).step(n).to_a.map! { |start| s[start, n] }
# => ["abcde", "fghij"]

Reply:
BTW, I would be a strong +1 for adding this as a method to more_core_extensions. I was thinking of the name `String#in_groups_of(n)` to mirror ActiveSupport's `in_groups_of` for Arrays since they have similar functionality.

Reply:
The blank line is actually intended, though, I will have to look to see what it ends up actually happening with my code.  But part of the "envelope" that `ansible-vault` puts this in does add the newline:

https://github.com/ansible/ansible/blob/c954c0727179064ce0ecf62336ec74173d36ec31/lib/ansible/parsing/vault/__init__.py#L236-L237

But whether I need the `.append("")` farther down here would be nice to determine, and I would remove it if it is already being added as part of the `.step` loop.

Reply:
I'm not saying the .append("") is incorrect...I'm saying the first 2 lines sometimes create a blank line and sometimes not, and then you tack on *another* blank line with the append, so you sometimes get 2 blanks and sometimes 1 blank.  For example:

```ruby
s = "abcedfghij"
n = 5
(0..s.size).step(n).to_a.map! { |start| s[start, n] }.append("").join("\n")
# => "abcde\nfghij\n\n"
n = 4
(0..s.size).step(n).to_a.map! { |start| s[start, n] }.append("").join("\n")
# => "abcd\nefgh\nij\n"
```

Making the range use `...` instead of `..` seems to fix it...


```ruby
s = "abcedfghij"
n = 5
(0...s.size).step(n).to_a.map! { |start| s[start, n] }.append("").join("\n")
# => "abcde\nfghij\n"
n = 4
(0...s.size).step(n).to_a.map! { |start| s[start, n] }.append("").join("\n")
# => "abcd\nefgh\nij\n"
```

Reply:
@Fryguy so I am getting completely different results from you...

```ruby
$ irb
irb(main):001:0> data = "123456789"
=> "123456789"
irb(main):002:0> (0..data.size).step(5).to_a.map!{ |start| data[start, 5] }
=> ["12345", "6789"]
irb(main):003:0> (0...data.size).step(5).to_a.map!{ |start| data[start, 5] }
=> ["12345", "6789"]
irb(main):004:0> exit
$ ruby -v
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-darwin18]
```

No empty string at the end... maybe it only happens when it mod divides evenly?

Reply:
Okay, yeah, if I change it `data = "1234567890"`, then I see the same.  I think I will change it to use `...` instead then, just so it is consistent.

Reply:
@Fryguy hehe, I think we typed this up at the same time... anyway, I agree, and I resolved.

Thanks!